packages/next/build/swc/src/transform.rs

```diff
@@ -0,0 +1,194 @@
+        complete_output(&env, result)
+    }
+}
+
+/// returns `compiler, (src / path), options, plugin, callback`
+pub fn schedule_transform<F>(cx: CallContext, op: F) -> napi::Result<JsObject>
+where
+    F: FnOnce(&Arc<Compiler>, String, bool, Options) -> TransformTask,
+{
+    let c = get_compiler(&cx);
+
+    let s = cx.get::<JsString>(0)?.into_utf8()?.as_str()?.to_owned();
+    let is_module = cx.get::<JsBoolean>(1)?;
+    let options: Options = cx.get_deserialized(2)?;
+
+    let task = op(&c, s, is_module.get_value()?, options);
+
+    cx.env.spawn(task).map(|t| t.promise_object())
+}
+
+pub fn exec_transform<F>(cx: CallContext, op: F) -> napi::Result<JsObject>
+where
+    F: FnOnce(&Compiler, String, &Options) -> Result<Arc<SourceFile>, Error>,
+{
+    let c = get_compiler(&cx);
+
+    let s = cx.get::<JsString>(0)?.into_utf8()?;
+    let is_module = cx.get::<JsBoolean>(1)?;
+    let options: Options = cx.get_deserialized(2)?;
+
+    let output = c.run(|| -> napi::Result<_> {
+        if is_module.get_value()? {
+            let program: Program =
+                serde_json::from_str(s.as_str()?).expect("failed to deserialize Program");
+            c.process_js(program, &options).convert_err()
+        } else {
+            let fm = op(&c, s.as_str()?.to_string(), &options).expect("failed to create fm");
+            c.process_js_file(fm, &options).convert_err()
+        }
+    })?;
+
+    complete_output(cx.env, output)
+}
+
+#[js_function(4)]
+pub fn transform(cx: CallContext) -> napi::Result<JsObject> {
+    schedule_transform(cx, |c, src, is_module, options| {
+        let input = if is_module {
+            Input::Program(src)
+        } else {
+            Input::Source(c.cm.new_source_file(
+                if options.filename.is_empty() {
+                    FileName::Anon
+                } else {
+                    FileName::Real(options.filename.clone().into())
+                },
+                src,
+            ))
+        };
+
+        TransformTask {
+            c: c.clone(),
+            input,
+            options,
+        }
+    })
+}
+
+#[js_function(4)]
+pub fn transform_sync(cx: CallContext) -> napi::Result<JsObject> {
+    exec_transform(cx, |c, src, options| {
+        Ok(c.cm.new_source_file(
+            if options.filename.is_empty() {
+                FileName::Anon
+            } else {
+                FileName::Real(options.filename.clone().into())
+            },
+            src,
+        ))
+    })
+}
+
+fn process_js_custom(
+    compiler: &Arc<Compiler>,
+    source: Arc<SourceFile>,
+    options: &Options,
+) -> Result<TransformOutput, Error> {
+    let config = compiler.run(|| compiler.config_for_file(options, &source.name))?;
+    let config = match config {
+        Some(v) => v,
+        None => {
+            bail!("cannot process file because it's ignored by .swcrc")
+        }
+    };
+    let config = BuiltConfig {
+        pass: chain!(hook_optimizer(), config.pass),
```

Comment:
Most of this code comes straight from swc except for this line which allows us to run our custom transforms before the builtin transforms.

Reply:
Yes, you can configure `target` to disable such transforms, but it's much better to use `swc_ecma_transforms` directly because it will completely remove `es3` passes from binary.

Reply:
Alright I will selectively choose what builtin transforms to use in a future PR.